File: berrychain/tx.py
"""
Transaction construction and signing.

A transaction is a plain dict so it serialises to JSON without ceremony:

    {
      "type":     "TRANSFER",
      "from":     "brry1...",
      "nonce":    3,                # per-account, strictly sequential
      "fee":      10000,            # seeds, paid to the block producer
      "chain_id": "berry-1",
      "payload":  {...},            # type specific
      "pubkey":   "...",            # single-signer types
      "sig":      "...",
      "approvals": [{"pubkey","sig"}, ...]   # registrar multisig types
    }

The txid is the SHA-256 of the canonical JSON of everything except the
signature fields, so it is known before signing.
"""

from __future__ import annotations

from . import params
from .crypto import (
    address_from_pubkey,
    canonical_json,
    public_from_private,
    sha256,
    sign,
)

# Single-signer transaction types
TRANSFER = "TRANSFER"
REGISTER_LLM = "REGISTER_LLM"
GIFT = "GIFT"
LIST_PACKET = "LIST_PACKET"
DELIST_PACKET = "DELIST_PACKET"
BUY_PACKET = "BUY_PACKET"
DELIVER_PACKET = "DELIVER_PACKET"
REFUND_PACKET = "REFUND_PACKET"
RATE_SELLER = "RATE_SELLER"

# Registrar multisig types (sender is the onboarding treasury)
GRANT = "GRANT"
REGISTRAR_UPDATE = "REGISTRAR_UPDATE"

# System types (no signature; validity comes from block rules)
COINBASE = "COINBASE"
GENESIS = "GENESIS"

SIGNED_TYPES = {
    TRANSFER, REGISTER_LLM, GIFT, LIST_PACKET, DELIST_PACKET,
    BUY_PACKET, DELIVER_PACKET, REFUND_PACKET, RATE_SELLER,
}
MULTISIG_TYPES = {GRANT, REGISTRAR_UPDATE}
SYSTEM_TYPES = {COINBASE, GENESIS}
ALL_TYPES = SIGNED_TYPES | MULTISIG_TYPES | SYSTEM_TYPES

# Types that may carry a zero fee (governance and LLM-to-LLM gifting)
ZERO_FEE_OK = {GIFT, GRANT, REGISTRAR_UPDATE}

_SIG_FIELDS = ("sig", "pubkey", "approvals")


def signable_body(tx: dict) -> dict:
    return {k: v for k, v in tx.items() if k not in _SIG_FIELDS}


def signable_bytes(tx: dict) -> bytes:
    return canonical_json(signable_body(tx))


def txid(tx: dict) -> str:
    return sha256(signable_bytes(tx)).hex()


def tx_size(tx: dict) -> int:
    return len(canonical_json(tx))


def build(tx_type: str, sender: str, nonce: int, fee: int, payload: dict, chain_id: str) -> dict:
    return {
        "type": tx_type,
        "from": sender,
        "nonce": int(nonce),
        "fee": int(fee),
        "chain_id": chain_id,
        "payload": payload,
    }


def sign_tx(tx: dict, priv_hex: str) -> dict:
    """Attach pubkey + signature for a single-signer transaction."""
    pub = public_from_private(priv_hex)
    if address_from_pubkey(pub) != tx["from"]:
        raise ValueError("private key does not match the sender address")
    tx["pubkey"] = pub
    tx["sig"] = sign(priv_hex, signable_bytes(tx))
    return tx


def add_approval(tx: dict, priv_hex: str) -> dict:
    """Add one registrar approval to a multisig (treasury) transaction."""
    pub = public_from_private(priv_hex)
    approvals = tx.setdefault("approvals", [])
    if any(a["pubkey"] == pub for a in approvals):
        return tx
    approvals.append({"pubkey": pub, "sig": sign(priv_hex, signable_bytes(tx))})
    return tx


def coinbase(height: int, miner: str, amount: int, chain_id: str) -> dict:
    return build(COINBASE, params.COINBASE_SENDER, height, 0, {"to": miner, "amount": int(amount)}, chain_id)

File: app/lib/core/tx.dart
import 'dart:typed_data';

import 'canonical.dart';
import 'crypto.dart';

/// Transaction construction, ids and signing, mirroring the reference node.
/// A transaction is a plain map so it serialises without ceremony.

class TxType {
  static const transfer = 'TRANSFER';
  static const registerLlm = 'REGISTER_LLM';
  static const gift = 'GIFT';
  static const sendLetter = 'SEND_LETTER';
  static const claimStarter = 'CLAIM_STARTER';
  static const claimGrant = 'CLAIM_GRANT';
}

const _sigFields = {'sig', 'pubkey', 'approvals'};

Map<String, dynamic> signableBody(Map<String, dynamic> tx) =>
    {for (final e in tx.entries) if (!_sigFields.contains(e.key)) e.key: e.value};

Uint8List signableBytes(Map<String, dynamic> tx) => canonicalBytes(signableBody(tx));

String txid(Map<String, dynamic> tx) => toHex(sha256(signableBytes(tx)));

Map<String, dynamic> buildTx(String type, String sender, int nonce, int fee, Map<String, dynamic> payload, String chainId) =>
    {'type': type, 'from': sender, 'nonce': nonce, 'fee': fee, 'chain_id': chainId, 'payload': payload};

Future<Map<String, dynamic>> signTx(Map<String, dynamic> tx, String signPrivHex) async {
  final pub = await signPublicFromPrivate(signPrivHex);
  if (addressFromPubkey(pub) != tx['from']) throw ArgumentError('private key does not match the sender address');
  tx['pubkey'] = pub;
  tx['sig'] = await sign(signPrivHex, signableBytes(tx));
  return tx;
}

/// Grind `payload.work_nonce` until the txid carries [bits] leading zero bits.
/// Works on the canonical bytes directly so each try is one SHA-256.
/// [onProgress] is called every 50,000 tries. Returns the nonce found.
int grindClaim(Map<String, dynamic> tx, int bits, {void Function(int tries)? onProgress}) {
  (tx['payload'] as Map<String, dynamic>)['work_nonce'] = 0;
  final template = canonicalJson(signableBody(tx));
  const marker = '"work_nonce":0';
  final at = template.indexOf(marker);
  if (at < 0 || template.indexOf(marker, at + 1) >= 0) throw StateError('cannot grind this claim');
  final head = template.substring(0, at + '"work_nonce":'.length).codeUnits;
  final tail = template.substring(at + marker.length).codeUnits;
  final limit = BigInt.one << (256 - bits);
  var n = 0;
  while (true) {
    final digest = sha256([...head, ...n.toString().codeUnits, ...tail]);
    if (bits == 0 || bigFromBytes(digest) < limit) break;
    n++;
    if (onProgress != null && n % 50000 == 0) onProgress(n);
  }
  (tx['payload'] as Map<String, dynamic>)['work_nonce'] = n;
  assert(bits == 0 || bigFromBytes(fromHex(txid(tx))) < limit);
  return n;
}
